File: Алгоритмизация/Практики/Практическая №7/1/main.cpp
#include <iostream>
#include <algorithm>
#include <climits>

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
    int maxSum;

    int maxGain(TreeNode* node) {
        if (node == nullptr) return 0;
        int leftGain = std::max(maxGain(node->left), 0);
        int rightGain = std::max(maxGain(node->right), 0);
        int priceNewPath = node->val + leftGain + rightGain;
        maxSum = std::max(maxSum, priceNewPath);
        return node->val + std::max(leftGain, rightGain);
    }

public:
    int maxPathSum(TreeNode* root) {
        maxSum = INT_MIN;
        maxGain(root);
        return maxSum;
    }
};

int main() {
    TreeNode* root = new TreeNode(-5);
    root->left = new TreeNode(1);
    root->right = new TreeNode(7);
    root->right->left = new TreeNode(2);
    root->right->right = new TreeNode(3);

    Solution sol;
    int result = sol.maxPathSum(root);
    std::cout << result << std::endl;

    return 0;
}

File: Алгоритмизация/Практики/Практическая №7/1/main.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def max_path_sum(self, root: TreeNode) -> int:
        self.max_sum = float('-inf')

        def max_gain(node):
            if not node:
                return 0
            left_gain = max(max_gain(node.left), 0)
            right_gain = max(max_gain(node.right), 0)
            price_new_path = node.val + left_gain + right_gain
            self.max_sum = max(self.max_sum, price_new_path)
            return node.val + max(left_gain, right_gain)

        max_gain(root)
        return self.max_sum

if __name__ == "__main__":
    root = TreeNode(-5)
    root.left = TreeNode(1)
    root.right = TreeNode(7)
    root.right.left = TreeNode(2)
    root.right.right = TreeNode(3)
    
    sol = Solution()
    print(sol.max_path_sum(root))
  
